src/Stratis.Bitcoin/BlockPulling/BlockPullerBehavior.cs

```diff
@@ -100,31 +121,71 @@ public BlockPullerBehavior(IBlockPuller blockPuller, IInitialBlockDownloadState
         }
 
         /// <inheritdoc/>
-        public void AddSample(double delaySeconds)
+        public void AddSample(long blockSizeBytes, double delaySinceRequestedSeconds)
         {
-            this.logger.LogTrace("({0}:{1})", nameof(delaySeconds), delaySeconds);
+            this.logger.LogTrace("({0}:{1},{2}:{3})", nameof(blockSizeBytes), blockSizeBytes, nameof(delaySinceRequestedSeconds), delaySinceRequestedSeconds);
+
+            double adjustedDelay;
+
+            if (this.lastDeliveryTime == null)
+                adjustedDelay = delaySinceRequestedSeconds;
+            else
+            {
+                double deliveryDiff = (DateTime.Now - this.lastDeliveryTime).Value.TotalSeconds;
+
+                adjustedDelay = Math.Min(delaySinceRequestedSeconds, deliveryDiff);
+            }
 
-            this.averageDelaySeconds.AddSample(delaySeconds);
+            this.lastDeliveryTime = DateTime.Now;
```

Comment:
Date time provider

Reply:
And probably move down below to be executed as late as possible

Reply:
Actually, maybe it would be better if this is set at the end of OnMessageReceivedAsync

Reply:
Otherwise penalise would also update it, do we want that? And also to avoid time for processing in the callback

Reply:
Don't wanna put it to OnMessageReceivedAsync because it will update in case we received a block we didn't ask for. 

Unless we change BP.PushBlock to return bool. 

Also overhead is not that bad when we penalize the peer. 

Let me know what you think

Reply:
Doesn't matter which block it is, we need to count it

Reply:
fixed